subscription/index.ts

```diff
@@ -0,0 +1,88 @@
+import { useEffect, useRef } from 'react'
+import useSWR from 'swr'
+import { withMiddleware } from '../src/utils/with-middleware'
+import { Key, SWRHook, Middleware, SWRConfiguration } from '../src/types'
+import { useIsomorphicLayoutEffect } from '../src/utils/env'
+
+export type SWRSubscription<Data = any, Error = any> = (
+  key: Key,
+  callback: (err?: Error, data?: Data) => void
+) => void
+
+export type SWRSubscriptionResponse<Data = any, Error = any> = {
+  data?: Data
+  error?: Error
+}
+
+export type SWRSubscriptionHook<Data = any, Error = any> = (
+  key: Key,
+  subscribe: SWRSubscription<Data, Error>,
+  config?: SWRConfiguration
+) => SWRSubscriptionResponse<Data, Error>
+
+const subscriptions = new Map<Key, number>()
+const disposers = new Map()
+
+export const subscription = (<Data, Error>(useSWRNext: SWRHook) =>
+  (
+    key: Key,
+    subscribe: SWRSubscription<Data, Error>,
+    config?: SWRConfiguration
+  ): SWRSubscriptionResponse<Data, Error> => {
+    const { data, error, mutate } = useSWRNext(key, null, config)
```

Comment:
We should not deconstruct the object here, but use `swr = useSWRNext(...)` and only access then when necessary. You can take other hooks for reference.